### PR TITLE
[node] Await waitUntil promises to resolve before exiting

### PR DESCRIPTION
**Pull request by nuta**
Say you have a middleware (`middleware.js`) that looks like this:

```js
export async default function middleware(req, ctx) {
  ctx.waitUntil(tooLongFunction())
}

async function tooLongFunction() {
  console.log('tooLongFunction started')
  await new Promise((resolve) => setTimeout(resolve, 10000))
  console.log('tooLongFunction finished')
}
```

When you run this middleware locally with `vercel dev`, you won't see the `tooLongFunction finished` message because the server process is killed right after finishing the response. While Edge Runtime's `server.close()`, which awaits all `waitUntil` promises, is called, but `exit-hook`'s callback in CLI doesn't wait for the `close` promise to resolve.

This PR fixes this by allowing an optional graceful shutdown callback instead of killing the process immediately.

### CODE REVIEW

**Review by craigandrews — DISMISSED** [review]

File: packages/node/package.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
     "edge-runtime": "2.5.7",
     "esbuild": "0.14.47",
     "etag": "1.8.1",
-    "exit-hook": "2.2.1",
+    "exit-hook": "4.0.0",
```
Comment (cb1kenobi) [inline packages/node/package.json]: I'm not sure we want to start introducing ESM dependencies quite yet. The CLI is still being transpiled to CommonJS and while `esbuild` does help with ESM compatibility, it hasn't been fully tested and there are known issues with dependencies that ESM features like `import.meta`.

Does `exit-hook@2.2.1` not work?

**Comment by cb1kenobi** [comment]
Can you please add a test?

File: packages/node/package.json (hunk 1 above)
Comment (Ethan-Arrowood) [reply packages/node/package.json]: agreed

Comment (TooTallNate) [reply packages/node/package.json]: I think this should be evaluated on a case by case basis. Seems like the tests are passing, so I'm fine with this update, especially if it's necessary to get the relevant hook for this change.

Comment (nuta) [reply packages/node/package.json]: > Does exit-hook@2.2.1 not work?

Unfortunately no as we need a newer version to use `asyncExitHook` API. I have another idea to fix this bug without upgrading this dependency. Will try it first 👍

**Comment by nuta** [comment]
Converted to draft until I finish rewriting this PR.

**Comment by nuta** [comment]
Rewrote the PR without using `exit-hook` package, with a test 👍 @cb1kenobi

**Review by Ethan-Arrowood — DISMISSED** [review]
very cool

**Review by Ethan-Arrowood — DISMISSED** [review]

**Review by EndangeredMassa — DISMISSED** [review]

**Comment by nuta** [comment]
^ the GitHub history got messed up but TL;DR: Fix merge conflicts in `pnpm-lock.yaml` and force-pushed my changes again.

**Review by EndangeredMassa — DISMISSED** [review]

**Comment by nuta** [comment]
@EndangeredMassa Can I get an approval again? 🙏 I've added a error handling (https://github.com/vercel/vercel/pull/10915/commits/35d839657f9a748871307f3a390fab7030994295) to handle uncaught exceptions in `waitUntil` promises.

**Review by EndangeredMassa — APPROVED** [review]

**Pull request merged**